### DISCUSSION TITLE
Implementation of pyenv and pipenv

### DISCUSSION THREAD

**Discussion by Mandroide** [ultralytics/yolov5#9279] [Ideas]
Hi. I think it would be a good idea to start using a Python Version Management Tool like [pyenv](https://github.com/pyenv/pyenv) for managing the python versions and possible virtual environments in the project, and [pipenv](https://pipenv.pypa.io/en/latest/index.html) for managing packages and its dependencies in the project.

I would like to know your opinion about it.

**Comment by glenn-jocher** [comment]
@Mandroide what would this mean in practice? i.e. exactly what would we do?

**Reply by Mandroide** [reply]
I mentioned pyenv because of the plain file ```.python-version``` that generates. This file could specify in the repo to which python version is the project, mainly aiming to avoid adding code that brings breaking changes. It is not required as such to have in the project, with the plain file specifying the python version is enough.

Pipenv generates Pipfile. This contains a snapshot of the precise set of packages that are installed, including direct dependencies as well as their _sub-dependencies_. With this approach, we wouldn't have to worry about keeping sub-dependencies up-to-date.

Pipenv also distinguishes between development dependencies and production (regular dependencies). For example, in the development dependencies we could define packages that are only needed in a dev environment and not for use of yolo like pytest or other that are not relevant to execute yolo. If you have used JavaScript before, this is like [npm](https://www.npmjs.com/).

The first time I saw those tools were from the article in [RealPython - Effective Python Environment](https://realpython.com/effective-python-environment/#package-management).

**Reply by glenn-jocher** [reply]
Ah, got it, thank you. We don’t have a single pip list since YOLOv5 runs in multiple environments like our docket images, CI runners and notebooks. You can see pip list outputs for the CI runners though in the ‘Check environments’ section of any YOLOv5 CI, ie most recent scheduled run from a few hours ago:
https://github.com/ultralytics/yolov5/runs/8218661895?check_suite_focus=true
